prog/dftb+/lib_dftb/encharges.F90

```diff
@@ -8,28 +8,442 @@
+
+    !> Instance of EEQ container
+    class(TEeqCont), intent(out) :: this
+
+    !> Input for the EEQ model
+    type(eeqInput), intent(in) :: input
+
+    !> Enable electrostatic energy calculation + derivatives
+    logical, intent(in) :: withEnergies
+
+    !> Enable partial charge calculation + derivatives
+    logical, intent(in) :: withCharges
+
+    !> Nr. of atoms in the system.
+    integer, intent(in) :: nAtom
+
+    !> Lattice vectors, if the system is periodic.
+    real(dp), intent(in), optional :: latVecs(:, :)
+
+    real(dp) :: recVecs(3, 3), maxGEwald
+
+    this%tPeriodic = present(latVecs)
+
+    this%param = input%TEeqParam
+    this%cutoff = input%cutoff
+    this%nrChrg = input%nrChrg
+
+    if (this%tPeriodic) then
+      this%tAutoEwald = input%parEwald <= 0.0_dp
+      this%tolEwald = input%tolEwald
+      this%parEwald = input%parEwald
+
+      call this%updateLatVecs(latVecs)
+    end if
+
+    this%nAtom = nAtom
+
+    if (withEnergies) then
+      allocate(this%energies(nAtom))
+      allocate(this%gradients(3, nAtom))
+      allocate(this%stress(3, 3))
+    end if
+
+    if (withCharges) then
+      allocate(this%charges(nAtom))
+      allocate(this%dqdr(3, nAtom, nAtom))
+      allocate(this%dqdL(3, 3, nAtom))
+    end if
+
+    this%tCoordsUpdated = .false.
+
+  end subroutine initialize
+
+
+  !> Notifies the objects about changed coordinates.
+  subroutine updateCoords(this, neigh, img2CentCell, coords, species, &
+      & cn, dcndr, dcndL)
+
+    !> Instance of EEQ container
+    class(TEeqCont), intent(inout) :: this
+
+    !> Updated neighbour list.
+    type(TNeighbourList), intent(in) :: neigh
+
+    !> Updated mapping to central cell.
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Updated coordinates.
+    real(dp), intent(in) :: coords(:,:)
+
+    !> Species of the atoms in the unit cell.
+    integer, intent(in) :: species(:)
+
+    real(dp), intent(in) :: cn(:), dcndr(:, :, :), dcndL(:, :, :)
+
+    integer, allocatable :: nNeigh(:)
+
+    allocate(nNeigh(this%nAtom))
+    call getNrOfNeighboursForAll(nNeigh, neigh, this%cutoff)
+
+    call getEEQCharges(this%nAtom, coords, species, this%nrChrg, nNeigh, &
+        & neigh%iNeighbour, neigh%neighDist2, img2CentCell, this%recPoint, this%parEwald, &
+        & this%vol, this%param%chi, this%param%kcn, this%param%gam, this%param%rad, cn, dcndr, dcndL, &
+        & this%energies, this%gradients, this%stress, &
+        & this%charges, this%dqdr, this%dqdL)
+
+    this%tCoordsUpdated = .true.
+
+  end subroutine updateCoords
+
+
+  !> Notifies the object about updated lattice vectors.
+  subroutine updateLatVecs(this, latVecs)
+
+    !> Instance of EEQ container
+    class(TEeqCont), intent(inout) :: this
+
+    !> New lattice vectors
+    real(dp), intent(in) :: latVecs(:, :)
+
+    real(dp) :: recVecs(3, 3), maxGEwald
+
+    @:ASSERT(this%tPeriodic)
+    @:ASSERT(all(shape(latVecs) == shape(this%latVecs)))
+
+    this%latVecs = latVecs
```

Comment:
this%latVecs(:,:) =
as a habit to avoid chance of doing this to an allocatable at some point.